### PR TITLE
ui-router

### PR DESCRIPTION
**Pull request by timkindberg**

### CODE REVIEW

File: lib/decorators/component.ts [hunk 1]
```diff
@@ -51,9 +51,17 @@ import directiveControllerFactory from '../util/directive-controller';
...
 
 const TYPE = 'component';
 
+export const componentHooks = {
+	after: [],
+	extendDDO: [],
+	beforeCtrlInvoke: [],
+	afterCtrlInvoke: []
+};
+
```
Comment (timkindberg) [inline lib/decorators/component.ts]: So the idea is to have various hooks in our Component's module parser. So far these are the ones I've thought might be needed. 
- after: happens after the whole parser has run, essentially if you just need to tack on some extra logic that isn't worried about intercepting or modify the DDO or ddo.controller.
- extendDDO: strictly for modifying the DDO before the call to angular.directive is made
- beforeCtrlInvoke: happens in the context of our controller 'hack' right before the real controller is invoked. Last chance to modify locals or other things.
- afterCtrlInvoke: Same as before but happens after Invoke.

**Comment by timkindberg** [comment]
@MikeRyan52 before I go any further I wanted to get feedback from you. Did you make any progress on a way to hook into / compose component modifications from external plugins?

File: lib/decorators/component.spec.ts [hunk 2]
```diff
@@ -377,6 +376,30 @@ describe('@Component', function(){
...
+			class MyClass{ }
+
+			let fixture = quickFixture({
+				directives: [MyClass],
+				template: `<foo>hello</foo>`
+			});
+
+			fixture.debugElement.text().should.be.equal('** hello **');
+		});
+
```
Comment (timkindberg) [inline lib/decorators/component.spec.ts]: These were just some missing tests... not part of this story

**Comment by timkindberg** [comment]
@MikeRyan52 I believe this is ready for official review / merge.

**Comment by timkindberg** [comment]
@brandonroberts can you review this and see if it meets your needs?

**Comment by brandonroberts** [comment]
@timkindberg Will do

File: lib/decorators/component.ts (hunk 1 above)
Comment (MikeRyan52) [inline lib/decorators/component.ts]: I like this. Seems simple and it is only meant to be a private API, right?

File: lib/decorators/component.ts [hunk 3]
```diff
@@ -192,50 +200,42 @@ export function View(
...
 Module.addProvider(TYPE, (target: any, name: string, injects: string[], ngModule: ng.IModule) => {
 	// First create an empty object to contain the directive definition object
 	let ddo: any = {};
 
-	componentStore.forEach((val, key) => {
-		// Loop through the key/val pairs of metadata and assign it to the DDO
-		ddo[key] = val;
-	}, target);
+	// Loop through the key/val pairs of metadata and assign it to the DDO
+	componentStore.forEach((val, key) => ddo[key] = val, target);
```
Comment (MikeRyan52) [inline lib/decorators/component.ts]: I know this is a piece of old code, but do you think this functionality should be moved into the Metastore class?

File: lib/decorators/component.ts (hunk 1 above)
Comment (MikeRyan52) [inline lib/decorators/component.ts]: I think we should wrap the function pushing mechanism with an interface that supports type checking. That way TS warns you if your function has the wrong call type signature.

Something like:
```ts
components.addAfterHook(function(target: any, name: string, injects: any, ngModule: angular.IModule){

});
```

**Comment by MikeRyan52** [comment]
Left some comments for minor improvements. Otherwise LGTM.

File: lib/decorators/component.ts (hunk 3 above)
Comment (timkindberg) [inline lib/decorators/component.ts]: I'm not sure how we'd move it there ??

File: lib/decorators/component.ts (hunk 1 above)
Comment (timkindberg) [inline lib/decorators/component.ts]: Good idea.

File: lib/decorators/state-config.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            throw new TypeError(createConfigErrorMessage(target, ngModule, '@StateConfig param must be an array of state objects.'));
+        }
+
+        ngModule.config(['$stateProvider', function($stateProvider: IStateProvider) {
+            if (!$stateProvider) return;
+
+            childStateConfigs.forEach((config: IComponentState) => {
+                // Grab tag name from component, use it to build a minimal state template
+                const tagName = providerStore.get('name', config.component);
+                config.template = config.template || `<${tagName}></${tagName}>`;
```
Comment (brandonroberts) [inline lib/decorators/state-config.ts]: The tagName should be dash case.

File: lib/decorators/state-config.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            componentStore.set(configsKey, [...existingConfigs, config], config.component);
+        });
+    }
+}
+
+function targetIsStaticFn(t) {
+    return t.name !== undefined && t.constructor.name === 'Function';
+}
+
+export function Resolve(resolveName: string = null){
```
Comment (brandonroberts) [inline lib/decorators/state-config.ts]: Can this be used to resolve promises?

Comment (brandonroberts) [inline lib/decorators/state-config.ts]: This also needs to be exposed in the main entry point

File: lib/decorators/state-config.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+const configsKey = 'ui-router.stateConfigs';
+const childConfigsKey = 'ui-router.stateChildConfigs';
+const annotatedResolvesKey = 'ui-router.annotatedResolves';
+const resolvedMapKey = 'ui-router.resolvedMap';
+
+export interface IComponentState extends IState {
+    component: any;
+}
+
+export function StateConfig(stateConfigs: IComponentState[]){
```
Comment (brandonroberts) [inline lib/decorators/state-config.ts]: This needs to be exposed in the main entry point

**Comment by brandonroberts** [comment]
I left a few comments. One issue/question I have deals with lazy loading of modules and may be outside the goal of this functionality. If you use multiple bundles, you won't be able to use this decorator as it requires you to import and define your components in your config. Is this mainly intended for a single bundle application?

**Comment by timkindberg** [comment]
Thanks @brandonroberts and @MikeRyan52 I'll review your comments and make the changes.

File: lib/decorators/state-config.ts (hunk 5 above)
Comment (timkindberg) [inline lib/decorators/state-config.ts]: Yes this function should work just like a regular resolve function so you should be able to return a promise.

File: lib/decorators/state-config.ts (hunk 4 above)
Comment (timkindberg) [inline lib/decorators/state-config.ts]: Great find. fixed.

**Comment by timkindberg** [comment]
@brandonroberts how would we get it to work with lazy loading?

**Comment by timkindberg** [comment]
So other than lazy loading, I made all the recommended changes. Thanks @MikeRyan52 and @brandonroberts for reviewing. I'm merging this.

**Comment by brandonroberts** [comment]
@timkindberg I think you would have to use "future states" from the ui-router extras. I think what you have is fine for most cases

**Comment by timkindberg** [comment]
@brandonroberts ok good to hear. Well what if we wanted to support lazy loaded routes out of the box? What might that look like? I've not used ui-router extras, I've only read the docs and stuff.

@christopherthielen any idea how hard it would be (and even how odd) for us to support lazy loaded states out of the box? Would it work well with our new @StateConfig and @Resolve decorators?

**Comment by brandonroberts** [comment]
@timkindberg instead of using the component function, you'd have to provide the component selector string and use that to provide the tag name in the template. It would also involve changing how the state configs provided by the decorator are stored.

**Pull request merged**